how-to/customize-workspace/client/src/connections.ts

```diff
@@ -0,0 +1,229 @@
+	console.log(`Action ${actionName} cleared from available actions list.`);
+	return true;
+}
+
+export async function getConnectedAppSourceClients() {
+	const connections: ConnectedAppSourceConnection[] = [];
+
+	const availableConnections = Object.values(connectedClients);
+	for (const connection of availableConnections) {
+		const matchedConnection = connection.connectionTypes.find(
+			(supportedConnectionType) => supportedConnectionType.type === "appSource"
+		);
+
+		if (matchedConnection !== undefined && matchedConnection.type === "appSource") {
+			connections.push({ identity: connection.identity, connectionData: matchedConnection });
+		}
+	}
+
+	return connections;
+}
+
+export async function getConnectedApps(): Promise<App[]> {
+	const connectedSources = await getConnectedAppSourceClients();
+	const apps: App[] = [];
+	for (let i = 0; i < connectedSources.length; i++) {
+		const returnedApplications: App[] = await connectionService.dispatch(
+			connectedSources[i].identity,
+			"getApps"
+		);
+		const manifestTypes = connectedSources[i]?.connectionData?.manifestTypes;
+		let validatedApps: App[] = [];
+		if (Array.isArray(manifestTypes) && manifestTypes.length > 0) {
+			validatedApps = returnedApplications.filter((entry) => manifestTypes.includes(entry.manifestType));
+		} else {
+			validatedApps = returnedApplications;
+		}
+		for (let v = 0; v < validatedApps.length; v++) {
+			if (validatedApps[v].manifestType === "connection") {
+				// ensure that if an app from a connection is marked as connection
+				// then it should only be sent to itself and not another uuid
+				validatedApps[v].manifest = connectedSources[i].identity.uuid;
+			}
+		}
+		apps.push(...validatedApps);
+	}
+	return apps;
+}
+
+export async function launchConnectedApp(app: App) {
+	const connectedSources = await getConnectedAppSourceClients();
+	const manifestType = "connection";
+	const connectedSource = connectedSources.find((entry) => entry.identity.uuid === app.manifest);
+	if (app.manifestType === manifestType && connectedSource !== undefined) {
+		console.log(`Launching app: ${app.appId} against connection: ${connectedSource.identity.uuid}`);
+		await connectionService.dispatch(connectedSource.identity, "launchApp", app);
+	} else {
+		console.warn(
+			`A request to launch app ${app.appId} was not successful. Either the manifestType is not ${manifestType}:${app.manifestType} or the connection ${app.manifest} is either not registered in the connectionProvider with appSource support or hasn't connected to this platform.`
+		);
+	}
+}
+
+export async function getConnectedSnapshotSourceClients() {
+	const connections: { identity: OpenFin.Identity; connectionData: SnapshotSourceConnection }[] = [];
+
+	const availableConnections = Object.values(connectedClients);
+	for (const connection of availableConnections) {
+		const matchedConnection = connection.connectionTypes.find(
+			(supportedConnectionType) => supportedConnectionType.type === "snapshotSource"
```

Comment:
The other const to go into shapes

Reply:
looked at it and that would make types have types and values (the string). I have created consts but kept them in the file to remove dupes of hardcoded strings and this is the only place they are referenced (if external parts need to reference them then we can move them out).